### ISSUE TITLE
Thread blocked and broker crash after more messages 

### ISSUE THREAD

**Issue by jalonsoa** [knative-sandbox/eventing-kafka-broker#1732]
**Describe the bug**

After tread blocked message we have more logs about session timeout, coordinator unavailable, consumer closed, etc .. and finaly a new exception and the broker stop.

```
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher Jan 12, 2022 1:55:32 PM io.vertx.core.impl.BlockedThreadChecker
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher WARNING: Thread Thread[vert.x-eventloop-thread-1,5,main] has been blocked for 13731 ms, time limit is 2000 ms
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher io.vertx.core.VertxException: Thread blocked
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at java.base@17.0.1/java.time.format.DateTimeFormatterBuilder$CompositePrinterParser.format(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at java.base@17.0.1/java.time.format.DateTimeFormatterBuilder$CompositePrinterParser.format(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at java.base@17.0.1/java.time.format.DateTimeFormatterBuilder$CompositePrinterParser.format(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at java.base@17.0.1/java.time.format.DateTimeFormatterBuilder$CompositePrinterParser.format(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at java.base@17.0.1/java.time.format.DateTimeFormatter.formatTo(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at java.base@17.0.1/java.time.format.DateTimeFormatter.format(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.cloudevents.types.Time.writeTime(Time.java:71)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.cloudevents.types.Time.writeTime(Time.java:84)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.cloudevents.rw.CloudEventContextWriter.withContextAttribute(CloudEventContextWriter.java:73)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.cloudevents.core.v1.CloudEventV1.readContext(CloudEventV1.java:191)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.cloudevents.core.impl.BaseCloudEvent.read(BaseCloudEvent.java:58)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.cloudevents.rw.CloudEventReader.read(CloudEventReader.java:38)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.cloudevents.core.message.MessageWriter.writeBinary(MessageWriter.java:71)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.http.WebClientCloudEventSender.send(WebClientCloudEventSender.java:83)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.http.WebClientCloudEventSender.lambda$send$0(WebClientCloudEventSender.java:72)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.http.WebClientCloudEventSender$$Lambda$848/0x00000008005558c8.handle(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.circuitbreaker.impl.CircuitBreakerImpl.executeOperation(CircuitBreakerImpl.java:382)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.circuitbreaker.impl.CircuitBreakerImpl.executeAndReportWithFallback(CircuitBreakerImpl.java:239)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.circuitbreaker.impl.CircuitBreakerImpl.executeWithFallback(CircuitBreakerImpl.java:398)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.circuitbreaker.impl.CircuitBreakerImpl.execute(CircuitBreakerImpl.java:403)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.http.WebClientCloudEventSender.send(WebClientCloudEventSender.java:70)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.RecordDispatcherImpl.lambda$composeSenderAndSinkHandler$8(RecordDispatcherImpl.java:230)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.RecordDispatcherImpl$$Lambda$470/0x000000080048e248.apply(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.RecordDispatcherImpl.onFilterMatching(RecordDispatcherImpl.java:171)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.RecordDispatcherImpl.onRecordReceived(RecordDispatcherImpl.java:163)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.RecordDispatcherImpl.dispatch(RecordDispatcherImpl.java:123)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.RecordDispatcherMutatorChain.dispatch(RecordDispatcherMutatorChain.java:58)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.consumer.UnorderedConsumerVerticle.lambda$startConsumer$0(UnorderedConsumerVerticle.java:35)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//dev.knative.eventing.kafka.broker.dispatcher.impl.consumer.UnorderedConsumerVerticle$$Lambda$481/0x000000080048c800.handle(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.kafka.client.consumer.impl.KafkaConsumerImpl.lambda$handler$1(KafkaConsumerImpl.java:81)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.kafka.client.consumer.impl.KafkaConsumerImpl$$Lambda$483/0x0000000800490270.handle(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.kafka.client.consumer.impl.KafkaReadStreamImpl.run(KafkaReadStreamImpl.java:237)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.kafka.client.consumer.impl.KafkaReadStreamImpl.lambda$schedule$8(KafkaReadStreamImpl.java:192)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.kafka.client.consumer.impl.KafkaReadStreamImpl$$Lambda$493/0x00000008004929c8.handle(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:63)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.core.impl.EventLoopContext.lambda$runOnContext$0(EventLoopContext.java:38)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.vertx.core.impl.EventLoopContext$$Lambda$243/0x0000000800381868.run(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 	at java.base@17.0.1/java.lang.Thread.run(Unknown Source)
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher 
```

```
kafka-broker-dispatcher-65c76bf5f-pc6s8 kafka-broker-dispatcher {"@timestamp":"2022-01-12T13:55:35.98Z","@version":"1","message":"An exception was thrown by io.vertx.core.net.impl.ConnectionBase$$Lambda$1033/0x00000008005de640.operationComplete()","logger_name":"io.netty.util.concurrent.DefaultPromise","thread_name":"vert.x-eventloop-thread-0","level":"WARN","level_value":30000,"stack_trace":"java.lang.IllegalStateException: complete already: DefaultChannelPromise@2b8ce4b7(failure: java.lang.IllegalStateException: Client is closed)\n\tat io.netty.util.concurrent.DefaultPromise.setSuccess(DefaultPromise.java:99)\n\tat io.netty.channel.DefaultChannelPromise.setSuccess(DefaultChannelPromise.java:78)\n\tat io.netty.channel.DefaultChannelPromise.setSuccess(DefaultChannelPromise.java:73)\n\tat io.vertx.core.net.impl.ConnectionBase.lambda$close$0(ConnectionBase.java:126)\n\tat io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:578)\n\tat io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:571)\n\tat io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:550)\n\tat io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:491)\n\tat io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:616)\n\tat io.netty.util.concurrent.DefaultPromise.setSuccess0(DefaultPromise.java:605)\n\tat io.netty.util.concurrent.DefaultPromise.setSuccess(DefaultPromise.java:96)\n\tat io.netty.channel.DefaultChannelPromise.setSuccess(DefaultChannelPromise.java:78)\n\tat io.netty.channel.DefaultChannelPromise.setSuccess(DefaultChannelPromise.java:73)\n\tat io.vertx.core.net.impl.ConnectionBase.handleClosed(ConnectionBase.java:381)\n\tat io.vertx.core.http.impl.Http1xClientConnection.handleClosed(Http1xClientConnection.java:1045)\n\tat io.vertx.core.net.impl.VertxHandler.channelInactive(VertxHandler.java:143)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241)\n\tat io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelInactive(CombinedChannelDuplexHandler.java:418)\n\tat io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:389)\n\tat io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:354)\n\tat io.netty.handler.codec.http.HttpClientCodec$Decoder.channelInactive(HttpClientCodec.java:326)\n\tat io.netty.channel.CombinedChannelDuplexHandler.channelInactive(CombinedChannelDuplexHandler.java:221)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241)\n\tat io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:81)\n\tat io.netty.handler.timeout.IdleStateHandler.channelInactive(IdleStateHandler.java:277)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)\n\tat io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:241)\n\tat io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1405)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:262)\n\tat io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:248)\n\tat io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:901)\n\tat io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:831)\n\tat io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)\n\tat io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasksFrom(SingleThreadEventExecutor.java:425)\n\tat io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:374)\n\tat io.netty.util.concurrent.SingleThreadEventExecutor.confirmShutdown(SingleThreadEventExecutor.java:760)\n\tat io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:527)\n\tat io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)\n\tat io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base/java.lang.Thread.run(Unknown Source)\n"}

```

 **Expected behavior**
No thread blocked error. The broker must continue and not stop.

 **To Reproduce**
Steps to reproduce the behavior.

 **Knative release version**

1.1.2

 **Additional context**

We have many triggers (+-150) ... sometimes some services can be slower or report many errors. May be related ?

**Comment by pierDipi** [comment]
Hi @jalonsoa, thanks for reporting!

Do you have pod metrics, CPU and memory usage of the dispatcher pod when this happens?

For a thread to be blocked for 13731 ms, there might be CPU throttling involved.

**Comment by jalonsoa** [comment]
Hi @pierDipi, thanks for the answer and for the fantastic work of the whole team !

Yes, we have problems with the CPU requests/limits ... but after solve this and avoid the cpu throttling, we continue with crash inside the dispatcher.

When the broker does not have accumulated messages, everything works perfectly, but if many messages accumulate in the broker (the group has a lot of lag: 300K by partition) and the service does not scale enough (timeout in the deliver of the services), the dispatchers go into a loop ( start -> open circuit -> thread blocks -> reconciling -> stop threads -> exceptions -> exit)

At this moment .. we have errors like:
```
kafka-broker-dispatcher-7f6585f985-ssf9x kafka-broker-dispatcher {"@timestamp":"2022-01-14T13:07:20.714Z","@version":"1","message":"Consumer exception","logger_name":"dev.knative.eventing.kafka.broker.dispatcher.impl.consumer.BaseConsumerVerticle","thread_name":"vert.x-kafka-consumer-thread-40","level":"ERROR","level_value":40000,"stack_trace":"java.lang.RuntimeException: java.lang.OutOfMemoryError: Java heap space\n\tat org.apache.kafka.clients.consumer.internals.AbstractCoordinator$HeartbeatThread.run(AbstractCoordinator.java:1453)\nCaused by: java.lang.OutOfMemoryError: Java heap space\n\tat java.base/java.nio.HeapByteBuffer.<init>(Unknown Source)\n\tat java.base/java.nio.ByteBuffer.allocate(Unknown Source)\n\tat org.apache.kafka.common.memory.MemoryPool$1.tryAllocate(MemoryPool.java:30)\n\tat org.apache.kafka.common.network.NetworkReceive.readFrom(NetworkReceive.java:113)\n\tat org.apache.kafka.common.network.KafkaChannel.receive(KafkaChannel.java:452)\n\tat org.apache.kafka.common.network.KafkaChannel.read(KafkaChannel.java:402)\n\tat org.apache.kafka.common.network.Selector.attemptRead(Selector.java:674)\n\tat org.apache.kafka.common.network.Selector.pollSelectionKeys(Selector.java:576)\n\tat org.apache.kafka.common.network.Selector.poll(Selector.java:481)\n\tat org.apache.kafka.clients.NetworkClient.poll(NetworkClient.java:561)\n\tat org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:265)\n\tat org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.pollNoWakeup(ConsumerNetworkClient.java:306)\n\tat org.apache.kafka.clients.consumer.internals.AbstractCoordinator$HeartbeatThread.run(AbstractCoordinator.java:1372)\n"}
```

**Comment by pierDipi** [comment]
Hi @jalonsoa, thanks for replying!

I'll try to reproduce the issue.

Would you mind sharing:

- the requests and limits (CPU and Memory) of the kafka-broker-dispatcher deployment
- the number of replicas
- min, max, and the average number of partitions of the topics

As a suggestion, since the error is `java.lang.OutOfMemoryError: Java heap space`, I'd try to tune a bit the [Consumer configurations](https://github.com/knative-sandbox/eventing-kafka-broker/blob/3f3f37ca49a64f6d3181b4296ce8c2b7b39e7028/data-plane/config/broker/100-config-kafka-broker-data-plane.yaml#L54), in particular, lowering the [fetch.max.bytes](https://github.com/knative-sandbox/eventing-kafka-broker/blob/3f3f37ca49a64f6d3181b4296ce8c2b7b39e7028/data-plane/config/broker/100-config-kafka-broker-data-plane.yaml#L73) and [max.poll.records](https://github.com/knative-sandbox/eventing-kafka-broker/blob/3f3f37ca49a64f6d3181b4296ce8c2b7b39e7028/data-plane/config/broker/100-config-kafka-broker-data-plane.yaml#L76) to lower the memory consumption.

We're improving the dispatcher scaling in https://github.com/knative-sandbox/eventing-kafka-broker/issues/1537 but that might a while.

**Comment by jalonsoa** [comment]
Hi !
  kafka-broker-dispatcher resources:  cpu limit(1500m) cpu request (1500m)
  3 replicas of dispatcher, with 152 triggers, with 12 partitions per topic

We lower the max.poll.records from 500 to 100, but don't works ... the container consumes 28GB at the out of memory exception

![image](https://user-images.githubusercontent.com/266372/149553229-f3d3097c-0cec-47bf-be62-41a0fdafed90.png)

If I undertand right ... the dispatcher instantiates 1 consumer per trigger per partition on rebalance group ??

152 triggers * 12 partition / 3 replicas = 608 consumers per dispatcher ....

If we lower the parttions and increase replicas:

152 triggers * 6 partitions / 6 replicas => 152 consumers per dispatcher ...

is ok ?

thanks !

**Comment by pierDipi** [comment]
It runs 1 consumer instance per dispatcher replica, so you would have 152 consumers per dispatcher.

I implemented back-pressure in https://github.com/knative-sandbox/eventing-kafka-broker/pull/1750 to reduce memory consumption in the case the service processing time is high.
(this week that PR will be released in 1.2.0)

https://github.com/knative-sandbox/eventing-kafka-broker/issues/1537 will come next, this helps your use case too.
